### PR TITLE
chore: migrate BaseInvertedFlatList.js to function component

### PR DESCRIPTION
**Pull request by teneeto**
### Details

This change is made to migrate the [BaseInvertedFlatList.js](https://github.com/Expensify/App/blob/main/src/components/InvertedFlatList/BaseInvertedFlatList.js) from class to a function component

### Fixed Issues

$: [16168](https://github.com/Expensify/App/issues/16168)
PROPOSAL:  [16168](https://github.com/Expensify/App/issues/16168)

### Tests

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://github.com/Expensify/App/assets/22779249/b19c3750-9b48-4ef5-b73a-09f96fbbf4b1

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://github.com/Expensify/App/assets/22779249/09957aa3-5b36-4c05-919b-05137402a7b2

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://github.com/Expensify/App/assets/22779249/e0d11ad5-a09f-4ecb-8cd4-27577321a7b9

</details>

<details>
<summary>Desktop</summary>

https://github.com/Expensify/App/assets/22779249/e96a6349-ffd9-4701-90f1-54c03b58bee7

</details>

<details>
<summary>iOS</summary>

https://github.com/Expensify/App/assets/22779249/a00353ab-390f-4bb2-befc-8c194f1d72fa

</details>

<details>
<summary>Android</summary>

https://github.com/Expensify/App/assets/22779249/e45c4000-5e6d-4924-9dd4-c47975ff05a8

</details>

### CODE REVIEW

File: src/components/InvertedFlatList/BaseInvertedFlatList.js
Comment (gedu) [inline src/components/InvertedFlatList/BaseInvertedFlatList.js:31]:
```suggestion
function BaseInvertedFlatList({initialRowHeight, shouldMeasureItems, innerRef, renderItem}) {
```

File: src/components/InvertedFlatList/BaseInvertedFlatList.js
Comment (gedu) [inline src/components/InvertedFlatList/BaseInvertedFlatList.js:77]: `sizeMap` shouldn't be in the dependencies because a ref won't produce a rerender, so if it changes a new function won't be created.

On the other hand, I don't see a need to memo this `getItemLayout`

File: src/components/InvertedFlatList/BaseInvertedFlatList.js
Comment (gedu) [inline src/components/InvertedFlatList/BaseInvertedFlatList.js:107]: `sizeMap` because it is a ref it won't produce a rerender when it changes, so you can remove it from dependencies

File: src/components/InvertedFlatList/BaseInvertedFlatList.js
Comment (teneeto) [reply src/components/InvertedFlatList/BaseInvertedFlatList.js:31]: Hi @gedu, I'm kind of aware of the props destructuring, but I had to leave it there for this case because of props forwarding and the forward ref that is used in this component.

File: src/components/InvertedFlatList/BaseInvertedFlatList.js
Comment (teneeto) [reply src/components/InvertedFlatList/BaseInvertedFlatList.js:77]: Honestly, I wasn't also convinced about using a memo here - but just left it to get your thoughts on it.

File: src/components/InvertedFlatList/BaseInvertedFlatList.js
Comment (teneeto) [reply src/components/InvertedFlatList/BaseInvertedFlatList.js:107]: @gedu If i remove sizeMap here, lint will definitely yell at it because ```measureItemLayout``` is a dependency in the ```renderItemFromProp```. Two things i could do here, i could disable lint warning - or entirely remove all memo. what do you suggest?

**Comment by teneeto** [comment]
Hi @cead22, any chance we could get this reviewed anytime soon?

File: src/components/InvertedFlatList/BaseInvertedFlatList.js
Comment (cead22) [inline src/components/InvertedFlatList/BaseInvertedFlatList.js:131]:
```suggestion
            renderItem={renderItemFromProp}
            
```

File: src/components/InvertedFlatList/BaseInvertedFlatList.js
Comment (cead22) [inline src/components/InvertedFlatList/BaseInvertedFlatList.js:134]:
```suggestion
            getItemLayout={shouldMeasureItems ? getItemLayout : undefined}
            
```

File: src/components/InvertedFlatList/BaseInvertedFlatList.js
Comment (cead22) [inline src/components/InvertedFlatList/BaseInvertedFlatList.js:141]: I know this comment already existed, but we shouldn't leave TODOs in the code, we should create GH issues for them instead. We have version control, so we also shouldn't leave code commented out
```suggestion

            // Commenting the line below as it breaks the unread indicator test
            // we will look at fixing/reusing this after RN v0.72
            // maintainVisibleContentPosition={{minIndexForVisible: 0, autoscrollToTopThreshold: 0}}
```

**Review by cead22 — CHANGES_REQUESTED** [review]

File: src/components/InvertedFlatList/BaseInvertedFlatList.js
Comment (teneeto) [reply src/components/InvertedFlatList/BaseInvertedFlatList.js:131]: This suggested commit would is causing lint to fail, so would revert change.

File: src/components/InvertedFlatList/BaseInvertedFlatList.js
Comment (teneeto) [reply src/components/InvertedFlatList/BaseInvertedFlatList.js:135]: Here also, the suggested commit is causing lint to fail, so would revert change.

**Comment by teneeto** [comment]
@cead22 thanks for the reviews, resolved all, you might want to take a look again.

**Review by cead22 — DISMISSED** [review]

**Comment by cead22** [comment]
@gedu do you mind giving this another look before I merge

File: src/components/InvertedFlatList/BaseInvertedFlatList.js
Comment (gedu) [inline src/components/InvertedFlatList/BaseInvertedFlatList.js:146]: Seems from this [thread](https://expensify.slack.com/archives/C01GTK53T8Q/p1695110739750909) that we are changing the way to use `forwardRef`, the documentation will be updated soon, I think we can use the new way

**Comment by cead22** [comment]
> @cead22 will be a new way to use `forwardRef` do you think we should add it here? or on a follow-up PR?

Can you share more details? I'm not familiar with this change. If we're ready to do it now and that's the ideal way to do it moving forward, then let's do it in this PR

**Comment by gedu** [comment]
@cead22 just double check and isn't ready yet, so I think the PR is good

**Comment by cead22** [comment]
@teneeto conflicts. @s77rt can you please complete the reviewer checklist? I'm not sure why you were automatically assigned to this (or @danieldoglas for that matter)

File: src/components/InvertedFlatList/BaseInvertedFlatList.js
Comment (s77rt) [inline src/components/InvertedFlatList/BaseInvertedFlatList.js:145]: NAB. This comment is missing in the new functional component

**Review by s77rt — CHANGES_REQUESTED** [review]

File: src/components/InvertedFlatList/BaseInvertedFlatList.js
Comment (cead22) [reply src/components/InvertedFlatList/BaseInvertedFlatList.js:145]: See https://github.com/Expensify/App/pull/26921#discussion_r1328274356

**Comment by s77rt** [comment]
## Reviewer Checklist

- [X] I have verified the author checklist is complete (all boxes are checked off).
- [X] I verified the correct issue is linked in the `### Fixed Issues` section above
- [X] I verified testing steps are clear and they cover the changes made in this PR
    - [X] I verified the steps for local testing are in the `Tests` section
    - [X] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [X] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [X] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [X] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [X] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android / native
    - [X] Android / Chrome
    - [ ] iOS / native
    - [X] iOS / Safari
    - [X] MacOS / Chrome / Safari
    - [X] MacOS / Desktop
- [X] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [X] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [X] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [X] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [X] I verified that comments were added to code that is not self explanatory
    - [X] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [X] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [X] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [X] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [X] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [X] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [X] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [X] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [X] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [X] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [X] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [X] If a new component is created I verified that:
    - [X] A similar component doesn't exist in the codebase
    - [X] All props are defined accurately and each prop has a `/** comment above it */`
    - [X] The file is named correctly
    - [X] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [X] The only data being stored in the state is data necessary for rendering and nothing else
    - [X] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [X] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [X] All JSX used for rendering exists in the render method
    - [X] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [X] If any new file was added I verified that:
    - [X] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [X] If a new CSS style is added I verified that:
    - [X] A similar style doesn't already exist
    - [X] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [X] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [X] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [X] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [X] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [X] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [X] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [ ] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>


web

</details>

<details>
<summary>Mobile Web - Chrome</summary>


mweb-chrome

</details>

<details>
<summary>Mobile Web - Safari</summary>


mweb-safari

</details>

<details>
<summary>Desktop</summary>


desktop

</details>

<details>
<summary>iOS</summary>



</details>

<details>
<summary>Android</summary>



</details>

**Review by s77rt — APPROVED** [review]

**Comment by danieldoglas** [comment]
My bad it took me a while to get to this. We have merge conflicts

**Comment by cead22** [comment]
@teneeto in case you missed the previous notification, we still have conflicts in this PR

**Comment by teneeto** [comment]
> @teneeto in case you missed the previous notification, we still have conflicts in this PR

Hi @cead22, sorry I was out for a while, but will attend to it ASAP.

**Comment by cead22** [comment]
No problem, and no rush, I just wanted to make sure you saw this

**Comment by teneeto** [comment]
@cead22 @s77rt Conflict is resolved. This might need your review and approval again. Thanks.

**Comment by s77rt** [comment]
Retested. Looks good to me 👍

**Comment by teneeto** [comment]
A fresh approval is needed here!

File: src/components/InvertedFlatList/BaseInvertedFlatList.js
Comment (cead22) [inline src/components/InvertedFlatList/BaseInvertedFlatList.js:128]: Is there a reason we're suppressing this here instead of updating the code to comply with this rule?

Comment (cead22) [reply src/components/InvertedFlatList/BaseInvertedFlatList.js:128]: I work on App much much less than you all, but I actually don't think ignoring the rule is better, and I'd love for us to consider passing all props explicitly
- It makes the code easier to understand, because you know what's being passed instead of going up the component hierarchy
- It doesn't allow us to send extra props that aren't needed (maybe there are ways this is prevented that I don't know about)
- Remove the rule would be better than having it and ignoring it

Comment (tgolen) [reply src/components/InvertedFlatList/BaseInvertedFlatList.js:128]: > When not to use the rule: If you are not using JSX or have lots of props to be passed or the props spreading is used inside HOC.

I think this is close, except the "have lots of props to be passed" is kind of misleading. The case where it's OK to use the spread operator is when you have a wrapper component that can pass _any vast number_ of props and there is no way to know what those props will be.

This is why it's OK to be used inside an HOC (because it's generic and any component can be passing props).

I think there is another case worth mentioning and that is when there is a wrapper around a native-built or a third-party component. I don't think we would want to try and explicitly pass every prop that is available to those components (ie. we might not know all the props, or the props might change with library updates that we would not catch and know about).

That's why I think there is a pretty good case for using the spread operator here.

For the majority of cases, we shouldn't use the spread operator and therefore it's still valuable for the ESlint rule to be there. It gives us a chance to do a thorough review of all the exceptions.

Comment (teneeto) [reply src/components/InvertedFlatList/BaseInvertedFlatList.js:128]: @cead22 how do we proceed with this? should we not suppress the warning? Also AFFAIK, there are many other exceptional cases as this for HOCs.

Comment (cead22) [reply src/components/InvertedFlatList/BaseInvertedFlatList.js:128]: Yeah that makes sense. We can continue that discussion elsewhere without blocking this PR on it

**Review by cead22 — APPROVED** [review]

**Pull request merged**
